File: src/engine/meld_analyzer.ts
import { Card, Rank, Suit, isChiChi } from '../models/card';
import { Hand, cardsOfRank } from '../models/hand';
import {
  Meld, MeldType, ChanMeld, CaMeld, BaDauMeld, QuanLe, MeldOrLe,
  createChan, createCa, createBaDau, countChan, isChan, isCa, isBaDau, isQuanLe,
} from '../models/meld';

export interface HandAnalysis {
  readonly chans: ChanMeld[];
  readonly cas: CaMeld[];
  readonly baDaus: BaDauMeld[];
  readonly quanLes: QuanLe[];
  readonly allMelds: MeldOrLe[];
  readonly chanCount: number;
}

// Analyze a hand and extract all melds
export function analyzeHand(hand: Hand): HandAnalysis {
  const cards = [...hand.cards];
  const chans: ChanMeld[] = [];
  const cas: CaMeld[] = [];
  const baDaus: BaDauMeld[] = [];
  const quanLes: QuanLe[] = [];

  const used = new Set<string>();

  // Group cards by rank
  const byRank = new Map<Rank, Card[]>();
  for (const card of cards) {
    if (!byRank.has(card.rank)) byRank.set(card.rank, []);
    byRank.get(card.rank)!.push(card);
  }

  // First pass: find Chăn (identical pairs)
  for (const [, rankCards] of byRank) {
    if (rankCards.length >= 2) {
      // Group by suit within this rank
      const bySuit = new Map<Suit, Card[]>();
      for (const card of rankCards) {
        if (!bySuit.has(card.suit)) bySuit.set(card.suit, []);
        bySuit.get(card.suit)!.push(card);
      }

      for (const [, suitCards] of bySuit) {
        while (suitCards.length >= 2) {
          const c1 = suitCards.pop()!;
          const c2 = suitCards.pop()!;
          chans.push(createChan(c1, c2));
          used.add(c1.id);
          used.add(c2.id);
        }
      }
    }
  }

  // Second pass: find Ba Đầu and Cạ from remaining cards
  const remainingByRank = new Map<Rank, Card[]>();
  for (const card of cards) {
    if (used.has(card.id)) continue;
    if (!remainingByRank.has(card.rank)) remainingByRank.set(card.rank, []);
    remainingByRank.get(card.rank)!.push(card);
  }

  for (const [, rankCards] of remainingByRank) {
    if (rankCards.length >= 3) {
      // Try to form Ba Đầu (all 3 suits present)
      const suits = rankCards.map(c => c.suit);
      const uniqueSuits = new Set(suits);

      if (uniqueSuits.size >= 3) {
        // Pick one card from each suit
        const picked: Card[] = [];
        const seenSuits = new Set<Suit>();
        for (const card of rankCards) {
          if (!seenSuits.has(card.suit) && picked.length < 3) {
            picked.push(card);
            seenSuits.add(card.suit);
          }
        }
        if (picked.length === 3) {
          baDaus.push(createBaDau(picked[0], picked[1], picked[2]));
          for (const card of picked) used.add(card.id);
        }
      }
    }

    if (rankCards.length === 2) {
      const remaining = rankCards.filter(c => !used.has(c.id));
      if (remaining.length === 2) {
        cas.push(createCa(remaining[0], remaining[1]));
        used.add(remaining[0].id);
        used.add(remaining[1].id);
      }
    }
  }

  // Third pass: remaining single cards are Què
  for (const card of cards) {
    if (!used.has(card.id)) {
      quanLes.push({ type: MeldType.QuanLe, card });
    }
  }

  const allMelds: MeldOrLe[] = [...chans, ...cas, ...baDaus, ...quanLes];

  return {
    chans,
    cas,
    baDaus,
    quanLes,
    allMelds,
    chanCount: chans.length,
  };
}

// Check if adding a card to hand would form a Chăn with any existing card
export function wouldFormChan(hand: Hand, card: Card): boolean {
  return hand.cards.some(
    c => c.rank === card.rank && c.suit === card.suit
  );
}

// Check if adding a card to hand would form a Cạ with any existing card
export function wouldFormCa(hand: Hand, card: Card): boolean {
  return hand.cards.some(
    c => c.rank === card.rank && c.suit !== card.suit
  );
}

// Check if player has 3 of a kind and could Chíu the 4th
export function canChiu(hand: Hand, card: Card): boolean {
  const matchingCards = hand.cards.filter(
    c => c.rank === card.rank && c.suit === card.suit
  );
  return matchingCards.length >= 3;
}

// Check if hand has exactly 5 Chăn + 4 Cạ + 1 orphan (Bạch Thủ pattern)
export function isBachThuPattern(analysis: HandAnalysis): boolean {
  return analysis.chanCount === 5 &&
    analysis.cas.length === 4 &&
    analysis.quanLes.length === 1;
}

// Check if hand has ≥6 Chăn (Ù Rộng eligible)
export function isUWideEligible(analysis: HandAnalysis): boolean {
  return analysis.chanCount >= 6;
}

File: src/ChanDanGian.Core/Engine/MeldAnalyzer.cs
namespace ChanDanGian.Core.Engine;

using ChanDanGian.Core.Models;

public record HandAnalysis(
    List<ChanMeld> Chans,
    List<CaMeld> Cas,
    List<BaDauMeld> BaDaus,
    List<QuanLe> QuanLes,
    List<IMeld> AllMelds,
    int ChanCount)
{
    public static HandAnalysis Empty => new([], [], [], [], [], 0);
}

public static class MeldAnalyzer
{
    public static HandAnalysis AnalyzeHand(Hand hand)
    {
        var cards = hand.Cards;
        var chans = new List<ChanMeld>();
        var cas = new List<CaMeld>();
        var baDaus = new List<BaDauMeld>();
        var quanLes = new List<QuanLe>();
        var used = new HashSet<string>();

        // Group by rank
        var byRank = cards.GroupBy(c => c.Rank)
            .ToDictionary(g => g.Key, g => g.ToList());

        // First pass: find Chan (identical pairs)
        foreach (var (_, rankCards) in byRank)
        {
            var bySuit = rankCards.GroupBy(c => c.Suit)
                .ToDictionary(g => g.Key, g => g.ToList());

            foreach (var (_, suitCards) in bySuit)
            {
                var remaining = new List<Card>(suitCards);
                while (remaining.Count >= 2)
                {
                    var c1 = remaining[^1]; remaining.RemoveAt(remaining.Count - 1);
                    var c2 = remaining[^1]; remaining.RemoveAt(remaining.Count - 1);
                    chans.Add(MeldFactory.CreateChan(c1, c2));
                    used.Add(c1.Id);
                    used.Add(c2.Id);
                }
            }
        }

        // Second pass: Ba Dau and Ca from remaining
        var remainingByRank = cards
            .Where(c => !used.Contains(c.Id))
            .GroupBy(c => c.Rank)
            .ToDictionary(g => g.Key, g => g.ToList());

        foreach (var (_, rankCards) in remainingByRank)
        {
            if (rankCards.Count >= 3)
            {
                var uniqueSuits = rankCards.Select(c => c.Suit).Distinct().ToList();
                if (uniqueSuits.Count >= 3)
                {
                    var picked = new List<Card>();
                    var seenSuits = new HashSet<Suit>();
                    foreach (var card in rankCards)
                    {
                        if (!seenSuits.Contains(card.Suit) && picked.Count < 3)
                        {
                            picked.Add(card);
                            seenSuits.Add(card.Suit);
                        }
                    }
                    if (picked.Count == 3)
                    {
                        baDaus.Add(MeldFactory.CreateBaDau(picked[0], picked[1], picked[2]));
                        foreach (var c in picked) used.Add(c.Id);
                    }
                }
            }

            var remaining = rankCards.Where(c => !used.Contains(c.Id)).ToList();
            if (remaining.Count == 2)
            {
                cas.Add(MeldFactory.CreateCa(remaining[0], remaining[1]));
                used.Add(remaining[0].Id);
                used.Add(remaining[1].Id);
            }
        }

        // Third pass: remaining singles are Quan Le
        foreach (var card in cards)
        {
            if (!used.Contains(card.Id))
                quanLes.Add(new QuanLe(card));
        }

        var allMelds = new List<IMeld>();
        allMelds.AddRange(chans);
        allMelds.AddRange(cas);
        allMelds.AddRange(baDaus);
        allMelds.AddRange(quanLes);

        return new HandAnalysis(chans, cas, baDaus, quanLes, allMelds, chans.Count);
    }

    public static bool WouldFormChan(Hand hand, Card card) =>
        hand.Cards.Any(c => c.Rank == card.Rank && c.Suit == card.Suit);

    public static bool WouldFormCa(Hand hand, Card card) =>
        hand.Cards.Any(c => c.Rank == card.Rank && c.Suit != card.Suit);

    public static bool CanChiu(Hand hand, Card card) =>
        hand.Cards.Count(c => c.Rank == card.Rank && c.Suit == card.Suit) >= 3;

    public static bool IsBachThuPattern(HandAnalysis analysis) =>
        analysis.ChanCount == 5 &&
        analysis.Cas.Count == 4 &&
        analysis.QuanLes.Count == 1;

    public static bool IsUWideEligible(HandAnalysis analysis) =>
        analysis.ChanCount >= 6;
}
